yield_plugin/src/yield_plugin.cpp

```diff
@@ -564,128 +594,175 @@ namespace yield_plugin
+    if (original_tp.trajectory_points.empty())
+    {
+      RCLCPP_ERROR(nh_->get_logger(), "Yield plugin received empty trajectory plan in update_traj_for_object");
+      throw std::invalid_argument("Yield plugin received empty trajectory plan in update_traj_for_object");
+    }
 
-      RCLCPP_DEBUG_STREAM(nh_->get_logger(),"object_downtrack: " << object_downtrack);
+    carma_planning_msgs::msg::TrajectoryPlan update_tpp_vector;
 
-      double dist_to_object = object_downtrack - vehicle_downtrack;
-      RCLCPP_DEBUG_STREAM(nh_->get_logger(),"dist_to_object: " << dist_to_object);
+    // Get earliest collision object
+    std::optional<std::pair<carma_perception_msgs::msg::ExternalObject, double>> earliest_collision_obj_pair = get_earliest_collision_object_and_time(original_tp, external_objects);
 
-      RCLCPP_DEBUG_STREAM(nh_->get_logger(),"object speed: " << earliest_collision_obj.value().velocity.twist.linear.x);
+    if (!earliest_collision_obj_pair)
+    {
+      RCLCPP_DEBUG(nh_->get_logger(),"No collision detected, so trajectory not modified.");
+      return original_tp;
+    }
 
-      // Distance from the original trajectory point to the lead vehicle/object
-      double dist_x = earliest_collision_obj.value().pose.pose.position.x - original_tp.trajectory_points[0].x;
-      double dist_y = earliest_collision_obj.value().pose.pose.position.y - original_tp.trajectory_points[0].y;
-      double x_lead = sqrt(dist_x*dist_x + dist_y*dist_y);
+    carma_perception_msgs::msg::ExternalObject earliest_collision_obj = earliest_collision_obj_pair.value().first;
+    double earliest_collision_time_in_seconds = earliest_collision_obj_pair.value().second;
 
-      // roadway object position
-      double gap_time = (x_lead - config_.x_gap)/initial_velocity;
+    // Issue (https://github.com/usdot-fhwa-stol/carma-platform/issues/2155): If the yield_plugin can detect if the roadway object is moving along the route,
+    // it is able to plan yielding much earlier and smoother using on_route_vehicle_collision_horizon_in_s.
 
-      double goal_velocity = earliest_collision_obj.value().velocity.twist.linear.x;
-      // determine the safety inter-vehicle gap based on speed
-      double safety_gap = std::max(goal_velocity * gap_time, config_.x_gap);
-      if (config_.enable_adjustable_gap)
-      {
-        // check if a digital_gap is available
-        double digital_gap = check_traj_for_digital_min_gap(original_tp);
-        RCLCPP_DEBUG_STREAM(nh_->get_logger(),"digital_gap: " << digital_gap);
-        // if a digital gap is available, it is replaced as safety gap
-        safety_gap = std::max(safety_gap, digital_gap);
-      }
-      // safety gap is implemented
-      double goal_pos = x_lead - safety_gap;
+    lanelet::BasicPoint2d point(original_tp.trajectory_points[0].x,original_tp.trajectory_points[0].y);
+    double vehicle_downtrack = wm_->routeTrackPos(point).downtrack;
```

Comment:
These can be marked `const`. We should apply this to all applicable.

Reply:
applied const